apps/src/gamelab/GameLab.js

```diff
@@ -826,38 +826,132 @@ GameLab.prototype.onP5ExecutionStarting = function () {
-
-  this.initInterpreter();
-  // And execute the interpreter for the first time:
-  if (this.JSInterpreter && this.JSInterpreter.initialized()) {
-    // Start executing the interpreter's global code as long as a setup() method
-    // was provided. If not, we will skip running any interpreted code in the
-    // preload phase and wait until the setup phase.
-    this.preloadInProgress = true;
-    if (this.globalCodeRunsDuringPreload) {
-      this.JSInterpreter.executeInterpreter(true);
-      this.interpreterStarted = true;
+  let isReadyNow = false;
+  Promise.all([
+      this.preloadAnimations_(),
+      this.runPreloadEventHandler_()
+  ]).then(() => {
+    isReadyNow = true;
+    this.gameLabP5.notifyPreloadPhaseComplete();
+  });
+  return isReadyNow;
+};
 
-      // In addition, execute the global function called preload()
-      if (this.eventHandlers.preload) {
-        this.eventHandlers.preload.apply(null);
-      }
+/**
+ * Wait for animations to be loaded into memory and ready to use, then pass
+ * those animations to P5 to be loaded into the engine as animations.
+ * @returns {Promise} which resolves once animations are in memory in the redux
+ *          store and we've started loading them into P5.
+ *          Loading to P5 is also an async process but it has its own internal
+ *          effect on the P5 preloadCount, so we don't need to track it here.
+ * @private
+ */
+GameLab.prototype.preloadAnimations_ = function () {
+  let store = this.studioApp_.reduxStore;
+  return new Promise(resolve => {
+    if (this.areAnimationsReady_()) {
+      resolve();
     } else {
-      if (this.eventHandlers.preload) {
-        this.log("WARNING: preload() was ignored because setup() was not provided");
-        this.eventHandlers.preload = null;
+      // Watch store changes until all the animations are ready.
+      const unsubscribe = store.subscribe(() => {
+        if (this.areAnimationsReady_()) {
+          unsubscribe();
+          resolve();
+        }
+      });
+    }
+  }).then(() => {
+    // Animations are ready - send them to p5 to be loaded into the engine.
+    this.gameLabP5.preloadAnimations(store.getState().animationList);
+  });
+};
+
+/**
+ * Check whether all animations in the project animation list have been loaded
+ * into memory and are ready to use.
+ * @returns {boolean}
+ * @private
+ */
+GameLab.prototype.areAnimationsReady_ = function () {
+  const animationList = this.studioApp_.reduxStore.getState().animationList;
+  return animationList.orderedKeys.every(key => animationList.propsByKey[key].loadedFromSource);
+};
+
+/**
+ * Run the preload event handler, and optionally global code, and report when
+ * it is done by resolving a returned Promise.
+ * @returns {Promise} Which will resolve immediately if there is no code to run,
+ *          otherwise will resolve when the preload handler has completed.
+ * @private
+ */
+GameLab.prototype.runPreloadEventHandler_ = function () {
```

Comment:
could I get a little more context for this? What's an example of the global code that it might run?